attestation-agent/attester/src/tsm_report/mod.rs

```diff
@@ -0,0 +1,148 @@
+// Copyright (c) 2024 Intel Corporation
+//
+// SPDX-License-Identifier: Apache-2.0
+//
+
+use log::error;
+use std::path::{Path, PathBuf};
+use tempfile::tempdir_in;
+use thiserror::Error;
+
+const TSM_REPORT_PATH: &str = "/sys/kernel/config/tsm/report";
+const TSM_REPORT_PROVIDER_TDX: &str = "tdx_guest";
+const TSM_REPORT_PROVIDER_SEV: &str = "sev_guest";
+
+#[derive(Error, Debug)]
+pub enum TsmReportError {
+    #[error("Failed to access TSM Report path")]
+    NoTsmReports,
+    #[error("Failed to create TSM Report path instance: {0}")]
+    Open(#[from] std::io::Error),
+    #[error("Failed to access TSM Report attribute: {0}: {1}")]
+    Access(&'static str, #[source] std::io::Error),
+    #[error("Failed to open TSM Report path for provider {0:?} ({1:?} found)")]
+    MissingProvider(TsmReportProvider, TsmReportProvider),
+    #[error("TSM Report inblob write conflict: generation {0}, expected 1")]
+    InblobConflict(u32),
+    #[error("Failed to generate TSM Report: missing inblob (len=0)")]
+    InblobLen,
+}
+
+#[derive(PartialEq, Debug)]
+pub enum TsmReportProvider {
+    Tdx,
+    Sev,
+    Unknown,
+}
+
+pub enum TsmReportData {
+    Tdx(Vec<u8>),
+    Sev(u8, Vec<u8>),
+}
+
+// TsmReportPath instance represents a unique path on ConfigFS
+// provided by the TSM_REPORT attestation ABI. Currently, each
+// instance is a one-shot attestation request and the path is
+// automatically removed when the instance goes out of scope.
+pub struct TsmReportPath {
+    path: PathBuf,
+}
+
+impl Drop for TsmReportPath {
+    fn drop(&mut self) {
+        let _ = std::fs::remove_dir(self.path.as_path())
+            .map_err(|e| log::error!("Failed to remove TSM Report directory: {}", e));
+    }
+}
+
+impl TsmReportPath {
+    pub fn new(wanted: TsmReportProvider) -> Result<Self, TsmReportError> {
+        if !Path::new(TSM_REPORT_PATH).exists() {
+            return Err(TsmReportError::NoTsmReports);
+        }
+
+        let p = tempdir_in(TSM_REPORT_PATH).map_err(TsmReportError::Open)?;
+
+        // Remove the Drop set by tempdir_in() since it errors on ConfigFS
+        // and leaks the created path. We implement our own Drop that removes the
+        // path (rmdir way) when TsmReportPath instance goes out of scope.
+        let path = p.into_path();
+
+        let provider = match std::fs::read_to_string(path.as_path().join("provider")) {
+            Ok(p) => match p.trim_matches('\n') {
+                TSM_REPORT_PROVIDER_TDX => TsmReportProvider::Tdx,
+                TSM_REPORT_PROVIDER_SEV => TsmReportProvider::Sev,
+                &_ => TsmReportProvider::Unknown,
+            },
```

Comment:
Might be more idiomatic to implement this as a to_string method on the TsmReportProvider struct. Also you can probably use the crate `strum` to avoid having these constants.

Reply:
fixed